crates/esplora/src/async_ext.rs

```diff
@@ -74,16 +86,119 @@ pub trait EsploraAsyncExt {
+        // not need updating. We just return the previous tip and use that as the point of
+        // agreement.
+        if let Some(prev_tip) = prev_tip.as_ref() {
+            if new_tip_height < prev_tip.height() {
+                return Ok(prev_tip.clone());
+            }
+        }
+
+        // Fetch new block IDs that are to be included in the update. This includes:
+        // 1. Atomically fetched most-recent blocks so we have a consistent view even during reorgs.
+        // 2. Heights the caller is interested in (as specified in `get_heights`).
+        let mut new_blocks = {
+            let heights = (0..=new_tip_height).rev();
+            let hashes = self
+                .get_blocks(Some(new_tip_height))
+                .await?
+                .into_iter()
+                .map(|b| b.id);
+
+            let mut new_blocks = heights.zip(hashes).collect::<BTreeMap<u32, BlockHash>>();
+
+            for height in get_heights {
+                // do not fetch blocks higher than known tip
+                if height > new_tip_height {
+                    continue;
+                }
+                if let btree_map::Entry::Vacant(entry) = new_blocks.entry(height) {
+                    let hash = self.get_block_hash(height).await?;
+                    entry.insert(hash);
+                }
+            }
+
+            new_blocks
+        };
+
+        // Determine the checkpoint to start building our update tip from.
+        let first_cp = match prev_tip {
+            Some(old_tip) => {
+                let old_tip_height = old_tip.height();
+                let mut earliest_agreement_cp = Option::<CheckPoint>::None;
+
+                for old_cp in old_tip.iter() {
+                    let old_block = old_cp.block_id();
+
+                    let new_hash = match new_blocks.entry(old_block.height) {
+                        btree_map::Entry::Vacant(entry) => *entry.insert(
+                            if old_tip_height - old_block.height >= ASSUME_FINAL_DEPTH {
+                                old_block.hash
+                            } else {
+                                self.get_block_hash(old_block.height).await?
+                            },
+                        ),
+                        btree_map::Entry::Occupied(entry) => *entry.get(),
+                    };
+
+                    // Since we may introduce blocks below the point of agreement, we cannot break
+                    // here unconditionally. We only break if we guarantee there are no new heights
+                    // below our current.
+                    if old_block.hash == new_hash {
+                        earliest_agreement_cp = Some(old_cp);
+
+                        let first_new_height = *new_blocks
+                            .keys()
+                            .next()
+                            .expect("must have atleast one new block");
+                        if first_new_height <= old_block.height {
```

Comment:
@evanlinjin do you mind expanding on why this condition is necessary? I'm not able to understand from the comment

Reply:
The comparison operator was actually the wrong way around so thank you for pointing out this line!

We are trying to find the checkpoint to build our update chain from. Finding a point of agreement is not enough if we wish to introduce blocks lower than the point of agreement (if we allow this, new blocks below point of agreement will NOT be included in the update).

Therefore, the rule to find our "checkpoint base" is: checkpoint must be a point of agreement && `height <= lowest_introduced_block_height`.

Reply:
And note that anything equal or lower to `ASSUME_FINAL_DEPTH` is considered a point of agreement.